# Mineração de Dados utilizando a metodologia CRISP-DM

## 1. Introdução

### O nome Crisp-DM é uma sigla para o inglês Cross Industry Standard Process for Data Mining, que em português significa algo como “Processo Padrão Inter-Indústrias para Mineração de Dados”. O objetivo dessa metodologia é desenvolver modelos a partir da análise de informações e dados de um negócio para prever futuras falhas e soluções.

### A metodologia Crisp-DM é dividida em seis etapas fundamentais, cada uma com suas particularidades e funcionalidades.

### As três primeiras etapas tem como objetivo a coleta e organização dos dados a serem analisados, envolvendo o entedimento do negócio, dados e a preparação dos dados.

<img src='https://miro.medium.com/v2/resize:fit:988/0*tA5OjppLK627FfFo' width=360>

###  Na imagem acima, é possível ver todas as etapas dessa metodologia, e algo interessante é que ela é um ciclo, então as etapas podem ser revistas e os processos podem ser melhorados mesmo depois de estabelecidos.

## 2. Entendimento do negócio

### Objetivo do projeto:  Analisar uma amostra de dados de usuários do Spotify, que contém informações sobre o que os usuário desse serviço de Streaming escutam em suas playlists, com o objetivo de encontrar associações. Por exemplo, se o usuário assiste a música/artista X, qual a chance dele também ouvir a música/artista Y?  

## 3. Entendimento dos dados

### Os dados serão obtidos através do site AiCrowd, onde é possível baixar o dataset (OBS: Os dados estão sendo disponibilizados devido a um desafio proposto, os dados não podem ser utilizados para fins comerciais, ver mais em https://www.aicrowd.com/challenges/spotify-million-playlist-dataset-challenge).

### Informações da base de dados fornecidas pelo Spotify

### Genero
 * Male: 45%
 * Female: 54%
 * Unspecified: 0.5%
 * Nonbinary: 0.5%

### Idade
 * 13-17:  10%
 * 18-24:  43%
 * 25-34:  31%
 * 35-44:   9%
 * 45-54:   4%
 * 55+:     3%

### País
 * US: 100%
 
### Estatísticas
 - number of playlists 1000000
 - number of tracks 66346428
 - number of unique tracks 2262292
 - number of unique albums 734684
 - number of unique artists 295860
 - number of unique titles 92944
 - number of playlists with descriptions 18760
 - number of unique normalized titles 17381
 - avg playlist length 66.346428

### Formato dos dados: JSON

### Dados estruturados: Sim

### Demonstração

Cada arquivo é um dicionário JSON com dois campos:
*info* e *playlists*.

### campo `info` 
Contém informações gerais sobre o arquivo:

   * **slice** - the range of slices that in in this particular file - such as 0-999
   * ***version*** -  - the current version of the MPD (which should be v1)
   * ***description*** - a description of the MPD
   * ***license*** - licensing info for the MPD
   * ***generated_on*** - a timestamp indicating when the slice was generated.

### campo `playlists` 
É uma lista que normalmente contém 1000 playlists, cada playlist é um dicionário que contém os seguintes campos:

* ***pid*** - integer - playlist id - the MPD ID of this playlist. This is an integer between 0 and 999,999.
* ***name*** - string - the name of the playlist 
* ***description*** - optional string - if present, the description given to the playlist.  Note that user-provided playlist descrptions are a relatively new feature of Spotify, so most playlists do not have descriptions.
* ***modified_at*** - seconds - timestamp (in seconds since the epoch) when this playlist was last updated. Times are rounded to midnight GMT of the date when the playlist was last updated.
* ***num_artists*** - the total number of unique artists for the tracks in the playlist.
* ***num_albums*** - the number of unique albums for the tracks in the playlist
* ***num_tracks*** - the number of tracks in the playlist
* ***num_followers*** - the number of followers this playlist had at the time the MPD was created. (Note that the follower count does not including the playlist creator)
* ***num_edits*** - the number of separate editing sessions. Tracks added in a two hour window are considered to be added in a single editing session.
* ***duration_ms*** - the total duration of all the tracks in the playlist (in milliseconds)
* ***collaborative*** -  boolean - if true, the playlist is a collaborative playlist. Multiple users may contribute tracks to a collaborative playlist.
* ***tracks*** - an array of information about each track in the playlist. Each element in the array is a dictionary with the following fields:
   * ***track_name*** - the name of the track
   * ***track_uri*** - the Spotify URI of the track
   * ***album_name*** - the name of the track's album
   * ***album_uri*** - the Spotify URI of the album
   * ***artist_name*** - the name of the track's primary artist
   * ***artist_uri*** - the Spotify URI of track's primary artist
   * ***duration_ms*** - the duration of the track in milliseconds
   * ***pos*** - the position of the track in the playlist (zero-based)

Exemplo de uma playlist no JSON:
  
        {
            "name": "musical",
            "collaborative": "false",
            "pid": 5,
            "modified_at": 1493424000,
            "num_albums": 7,
            "num_tracks": 12,
            "num_followers": 1,
            "num_edits": 2,
            "duration_ms": 2657366,
            "num_artists": 6,
            "tracks": [
                {
                    "pos": 0,
                    "artist_name": "Degiheugi",
                    "track_uri": "spotify:track:7vqa3sDmtEaVJ2gcvxtRID",
                    "artist_uri": "spotify:artist:3V2paBXEoZIAhfZRJmo2jL",
                    "track_name": "Finalement",
                    "album_uri": "spotify:album:2KrRMJ9z7Xjoz1Az4O6UML",
                    "duration_ms": 166264,
                    "album_name": "Dancing Chords and Fireflies"
                },
                {
                    "pos": 1,
                    "artist_name": "Degiheugi",
                    "track_uri": "spotify:track:23EOmJivOZ88WJPUbIPjh6",
                    "artist_uri": "spotify:artist:3V2paBXEoZIAhfZRJmo2jL",
                    "track_name": "Betty",
                    "album_uri": "spotify:album:3lUSlvjUoHNA8IkNTqURqd",
                    "duration_ms": 235534,
                    "album_name": "Endless Smile"
                },
                {
                    "pos": 2,
                    "artist_name": "Degiheugi",
                    "track_uri": "spotify:track:1vaffTCJxkyqeJY7zF9a55",
                    "artist_uri": "spotify:artist:3V2paBXEoZIAhfZRJmo2jL",
                    "track_name": "Some Beat in My Head",
                    "album_uri": "spotify:album:2KrRMJ9z7Xjoz1Az4O6UML",
                    "duration_ms": 268050,
                    "album_name": "Dancing Chords and Fireflies"
                },
                // 8 tracks omitted
                {
                    "pos": 11,
                    "artist_name": "Mo' Horizons",
                    "track_uri": "spotify:track:7iwx00eBzeSSSy6xfESyWN",
                    "artist_uri": "spotify:artist:3tuX54dqgS8LsGUvNzgrpP",
                    "track_name": "Fever 99\u00b0",
                    "album_uri": "spotify:album:2Fg1t2tyOSGWkVYHlFfXVf",
                    "duration_ms": 364320,
                    "album_name": "Come Touch The Sun"
                }
            ],

        }


## 4. Preparação dos dados

### Os dados que serão utilizados já estão bem estruturados e documentados, o que diminui os esforços do processo de preparação e tratamento de dados. Porém algumas coisas devem ser feitas:

#### 1. Eliminar nulos e playlists sem músicas
#### 2. Reordernar dados para a estrutura csv, onde cada linha será uma playlist, exemplo: 
ID&emsp;&emsp;&emsp;, Item 1&emsp;&emsp;&emsp;&emsp;&emsp;, Item 2&emsp;&emsp;&emsp;&emsp;&emsp;, Item 3 &emsp;&emsp;&emsp;&emsp;&emsp;, Item 4&emsp;&emsp;&emsp;&emsp;&emsp;... <br>
1048588, artista X/musica Y, artista X/musica Z, artista X/musica Z, artista X/musica Z... <br>
1048589, artista X/musica Z, artista X/musica Z, artista X/musica Z, artista X/musica Z... <br>

### Visualização dos dados coletados:

In [2]:
import pandas as pd
import os
import numpy as np
import json
import random


In [27]:
arquivo = "./spotify_million_playlist_dataset/mpd.slice.843000-843999.json"
with open(arquivo, 'r') as file:
    data = json.load(file)

# Extrair a primeira playlist
primeira_playlist = data['playlists']

# Ler os dados diretamente como um DataFrame do pandas
viz = pd.json_normalize(primeira_playlist)

# Visualizar 10 primeiras linhas
viz[viz.num_tracks >0]

name collaborative     pid  modified_at  num_tracks  \
0           get hype         false  843000   1507161600          96   
1    party playlist          false  843001   1499126400         136   
2              d-.-b         false  843002   1473897600         237   
3              sleep         false  843003   1508457600          26   
4               1985         false  843004   1478476800          37   
..               ...           ...     ...          ...         ...   
995             main         false  843995   1507939200          45   
996               ⛵️         false  843996   1506384000          89   
997              Boi         false  843997   1505865600          24   
998          dubstep         false  843998   1486944000          14   
999            Go Go         false  843999   1446768000           9   

     num_albums  num_followers  \
0            72              2   
1           105              1   
2           192              1   
3            22              1   
4            30              1   
..          ...            ...   
995          42              1   
996          72              1   
997          24              1   
998          11              1   
999           9              1   

                                                tracks  num_edits  \
0    [{'pos': 0, 'artist_name': 'DRAM', 'track_uri'...          6   
1    [{'pos': 0, 'artist_name': 'Sonny Alven', 'tra...          3   
2    [{'pos': 0, 'artist_name': 'Pac Div', 'track_u...        126   
3    [{'pos': 0, 'artist_name': 'Drake', 'track_uri...          5   
4    [{'pos': 0, 'artist_name': 'The Cars', 'track_...          5   
..                                                 ...        ...   
995  [{'pos': 0, 'artist_name': 'The All-American R...         25   
996  [{'pos': 0, 'artist_name': 'Cage The Elephant'...         27   
997  [{'pos': 0, 'artist_name': 'Clairo', 'track_ur...          2   
998  [{'pos': 0, 'artist_name': 'Doctor P', 'track_...          7   
999  [{'pos': 0, 'artist_name': 'Northeast Groovers...          6   

     duration_ms  num_artists description  
0       22121466           43         NaN  
1       31381784           74         NaN  
2       57718299          121         NaN  
3        6843516           13         NaN  
4        9516934           24         NaN  
..           ...          ...         ...  
995      9974073           40         NaN  
996     20134760           49         NaN  
997      5902167           24         NaN  
998      3285193            9         NaN  
999      3881998            5         NaN  

[1000 rows x 12 columns]

### Os dados atuais possuem diversas informações que não serão utilizadas no modelo de associação, sendo assim, é necessário eliminar o excesso de informações.

In [26]:
# O Código irá percorrer cada arquivo JSON na pasta "spotify_million_playlist_dataset", e então obter apenas o campo "playlists"

# Isso irá diminuir o tamanho do arquivo final, e facilita ao selecionar uma amostra 
# aleatória, pois o dataset já estará dividido em grupos.


playlists = []
split_playlists = []

# Diretório onde estão os arquivos JSON, somando todos os arquivos, temos 8.1GB de dados disponíveis
diretorio = "./spotify_million_playlist_dataset/"
count = 0

# Iterar sobre todos os arquivos no diretório
for arq in os.listdir(diretorio):
    if arq.endswith(".json"):
        # Carregar o arquivo JSON
        with open(os.path.join(diretorio, arq), 'r') as f:
            data = json.load(f)
            # Extrair o campo "playlists" e adicionar à lista
            all_playlists.extend(data['playlists'])
            count = count + 1
            if count> 0 and count % 10 == 0:
                split_playlists.append(playlists)
                with open(f"all_playlists_{count}.json", 'w') as f:
                    json.dump({'playlists': split_playlists}, f)
                
                all_playlists =[]
                split_playlists=[]

# O processo demorou cerca de 11 minutos, sendo que cada item ficou com cerca de 196MB.
# foram gerados 28 arquivos.
# Somando todos os arquivos, temos 5,21GB, uma economia de quase 3GB.

### A próxima etapa é selecionar uma amostra, de 28 arquivos, serão selecionados 10 de forma totalmente aleatória

In [3]:
random_nums = list(range(10,290,10))
random.shuffle(random_nums)
# print(random_nums)
numeros_selecionados = []

for i in range(0,len(random_nums)):
    if i <= 10:
        numeros_selecionados.append(random_nums[i])

In [4]:
numeros_selecionados

[100, 120, 200, 270, 30, 60, 280, 140, 170, 40, 110]

### O código seguinte passa por cada arquivo selecionado aleatoriamente e grava o conteúdo em um arquivo final com todos os dados necessário

In [6]:
all_playlists = []

# Diretório onde estão os arquivos JSON
diretorio = "./"

# Iterar sobre todos os arquivos no diretório
for arq in os.listdir(diretorio):
    for num in numeros_selecionados:
        if arq.endswith(".json") and '_' + str(num) in arq:
            # Carregar o arquivo JSON
            with open(os.path.join(diretorio, arq), 'r') as f:
                data = json.load(f)
                # Extrair o campo "playlists" e adicionar à lista
                all_playlists.extend(data['playlists'])
                f.close()

# Criar um novo arquivo JSON com todas as playlists
with open('all_data.json', 'w') as f:
    json.dump({'playlists': all_playlists}, f)
    f.close()

In [4]:
arquivo = "all_data.json"
with open(arquivo, 'r') as file:
    data = json.load(file)
    file.close()

# Extrair as playlists
playlists = data['playlists']

dataframes = [pd.json_normalize(playlist) for playlist in playlists]

dados = pd.concat(dataframes, ignore_index=True)

dados['tracks'] = dados['tracks'].apply(lambda x: [f"{track['artist_name']} / {track['track_name']}" for track in x])
dados.head(10)


name collaborative     pid  modified_at  num_tracks  \
0                     2016         false  823000   1509235200          77   
1                    Fav❤️         false  823001   1503792000          99   
2              Christmas.          false  823002   1449273600          16   
3                 Eargasms         false  823003   1500940800          29   
4  Noosa Tastiest Playlist         false  823004   1472774400          33   
5                      ♡♡♡         false  823005   1493856000          35   
6                Christmas         false  823006   1481328000          42   
7              This Summer         false  823007   1473120000          40   
8                   maddie         false  823008   1509408000          85   
9                 Variado          false  823009   1476576000          33   

   num_albums  num_followers  \
0          76              1   
1          89              1   
2           9              1   
3          28              1   
4          33              1   
5          33              2   
6          31              2   
7          35              1   
8          74              1   
9          16              1   

                                              tracks  num_edits  duration_ms  \
0  [Jon Bellion / All Time Low, Tatiana Manaois /...         17     16680321   
1  [Petit Biscuit / Sunset Lover, Jai Wolf / Indi...         51     22127129   
2  [Michael Bublé / I'll Be Home For Christmas, M...          3      3055529   
3  [Dillon Francis / Coming Over (feat. James Her...         21      6292781   
4  [Lee Brice / Hard To Love, Dustin Lynch / Hell...          2      7160167   
5  [Train / Drops of Jupiter, Plain White T's / H...         11      8030962   
6  [Kenny Loggins / Celebrate Me Home, Josh Groba...          8      9132652   
7  [Squeeze / This Summer, Ra Ra Riot / Water, Sl...         28      8904808   
8  [Drake / One Dance, Rihanna / Work, Drake / Ju...         48     19047362   
9  [Francisco Céspedes / Todo es un misterio, Fra...          6      8251439   

   num_artists description  
0           75         NaN  
1           78         NaN  
2           12         NaN  
3           25         NaN  
4           33         NaN  
5           32         NaN  
6           29         NaN  
7           28         NaN  
8           62         NaN  
9           13         NaN

### Agora que os dados de amostra estão completos, é necessário analisar

In [5]:
dados.describe()

pid   modified_at     num_tracks     num_albums  \
count  110000.000000  1.100000e+05  110000.000000  110000.000000   
mean   791135.863636  1.476293e+09      66.455127      49.638500   
std    233208.510047  3.673033e+07      53.677038      39.911424   
min     77000.000000  1.275437e+09       5.000000       2.000000   
25%    777499.750000  1.458950e+09      26.000000      20.000000   
50%    839999.500000  1.490227e+09      49.000000      37.000000   
75%    895499.250000  1.505520e+09      92.000000      68.000000   
max    994999.000000  1.509494e+09     376.000000     242.000000   

       num_followers      num_edits   duration_ms    num_artists  
count  110000.000000  110000.000000  1.100000e+05  110000.000000  
mean        2.342927      17.722382  1.558836e+07      38.088755  
std        66.638100      20.659587  1.281889e+07      30.149510  
min         1.000000       1.000000  4.748380e+05       3.000000  
25%         1.000000       5.000000  6.012931e+06      16.000000  
50%         1.000000      10.000000  1.144368e+07      29.000000  
75%         1.000000      22.000000  2.145740e+07      52.000000  
max     15123.000000     197.000000  1.678909e+08     222.000000

### Usando apenas a análise estatística, é possível observar alguns pontos
 - A média de músicas por playlist é de +- 66
 - Possibilidade de outliers
 - As playlists estão desbalanceadas, algumas possuem 30 músicas, enquanto outras possuem 300. Isso não é bom para o modelo de associação.

### Removendo outliers


In [6]:
# Calcular o Z-score para cada ponto de dados
dados['Z_score'] = (dados['num_tracks'] - dados['num_tracks'].mean()) / dados['num_tracks'].std()

# Identificar outliers com Z-score acima de 3 ou abaixo de -3
outliers_z_score_index = dados[(dados['Z_score'] > 3) | (dados['Z_score'] < -3)].index

# Eliminar outliers
dados_sem_outliers = dados.drop(outliers_z_score_index)
len(dados_sem_outliers)

108591

### Preencher dados vazios com o valor Nan, facilitando o filtro.

In [7]:
tracks = []

for index, row in dados_sem_outliers.iterrows():
    tracks.append(row['tracks'])

new_df = pd.DataFrame(tracks)

new_df = new_df.fillna(np.nan)

new_df.columns = [f'Track {i}' for i in range(1, len(new_df.columns) + 1)]

In [8]:
new_df.tail(30)

Track 1  \
108561           Adele / Send My Love (To Your New Lover)   
108562                         Sara Bareilles / Uncharted   
108563                   Boys Like Girls / Hero / Heroine   
108564                              BØRNS / Electric Love   
108565                                   San Holo / Light   
108566                              FKA twigs / Two Weeks   
108567                    Miley Cyrus / Hoedown Throwdown   
108568  Hillsong Young & Free / Where You Are - Radio ...   
108569                          Washed Out / Great Escape   
108570                  Dustin Lynch / Cowboys and Angels   
108571                             Tritonal / Untouchable   
108572                            The Knife / The Captain   
108573                        Florida Georgia Line / Stay   
108574               Selena Gomez / Kill Em With Kindness   
108575                                Saje / Lost Tonight   
108576                              Kevin Gates / What If   
108577                         Alice Cooper / I Never Cry   
108578                         Childish Gambino / Redbone   
108579                       Petit Biscuit / Sunset Lover   
108580          Edward Sharpe & The Magnetic Zeros / Home   
108581                               alt-J / Fitzpleasure   
108582                               Drake / Marvins Room   
108583                    Portugal. The Man / So American   
108584                             Lil Wayne / Believe Me   
108585  Faith Hill / Where Are You Christmas - From "D...   
108586          Pentatonix / Somebody That I Used to Know   
108587                      Donnie McClurkin / The Prayer   
108588                        Trevor Hall / The Lime Tree   
108589        Major Lazer / Lean On (feat. MØ & DJ Snake)   
108590                   Newsboys / Your Love Never Fails   

                                                  Track 2  \
108561                                   Drake / Too Good   
108562                         Sara Bareilles / Manhattan   
108563                 Boys Like Girls / The Great Escape   
108564                             Laura Veirs / Sun Song   
108565                               Train / Working Girl   
108566                           Pentatonix / La La Latch   
108567                            Miley Cyrus / The Climb   
108568               Hillsong Young & Free / Sinking Deep   
108569                  Courtney Barnett / Avant Gardener   
108570                           John Waite / Missing You   
108571                              Jonas Blue / Fast Car   
108572  The Queen of Hearts / Where Are You Now? - Sta...   
108573                      Florida Georgia Line / Cruise   
108574                            Justin Bieber / Company   
108575                                  Flume / Sleepless   
108576                                 Kid Ink / F With U   
108577                 Christopher Cross / Think Of Laura   
108578      Cash Cash / All My Love (feat. Conor Maynard)   
108579                                   Wingtip / Rewind   
108580          The Cinematic Orchestra / To Build A Home   
108581  Cage The Elephant / Ain't No Rest for the Wick...   
108582                                   Drake / Too Good   
108583                        Portugal. The Man / The Sun   
108584                                Lil Wayne / A Milli   
108585                         Céline Dion / O Holy Night   
108586                    Pentatonix / Love You Long Time   
108587               Kirk Franklin / Praise Joint (Remix)   
108588                  James Vincent McMorrow / Red Dust   
108589                    The Weeknd / Can't Feel My Face   
108590        Family Force 5 / Cray Button (feat. Lecrae)   

                                                  Track 3  \
108561                           Ariana Grande / Into You   
108562                           Sara Bareilles / Gravity   
108563                          Boys Like Girls / Thunder   
108564                

### Verificando os valores nulos, é possível ver o número de nulos aumenta conforme o tamanho da playlist, como temos 108591 registros, podemos definir um número de músicas por playlist que mantenha pelo menos metade desses registros, eliminando os dados nulos da análise. 

### Poderia ser definido que todos os registros teriam apenas 5 musicas, pois não existe nenhum registro que não tenha pelo menos esse número de músicas

### Porém, é necessário um equilíbrio dos dados, pois com poucas informações sobre a playlist, o algoritmo de associação não vai ter um resultado tão bom.

### Sendo assim, será definido que o número de músicas em uma playlist será 66 (média de músicas por playlist), e com isso irá sobrar 40.356 dados para análise.

In [9]:
pd.set_option('display.max_rows', None)
new_df.isna().sum()


Track 1           0
Track 2           0
Track 3           0
Track 4           0
Track 5           0
Track 6         600
Track 7        1287
Track 8        2037
Track 9        2968
Track 10       3937
Track 11       5075
Track 12       6258
Track 13       7621
Track 14       8932
Track 15      10382
Track 16      11898
Track 17      13401
Track 18      14885
Track 19      16374
Track 20      17877
Track 21      19514
Track 22      21128
Track 23      22555
Track 24      23984
Track 25      25438
Track 26      26899
Track 27      28325
Track 28      29680
Track 29      31026
Track 30      32336
Track 31      33808
Track 32      35122
Track 33      36422
Track 34      37690
Track 35      38891
Track 36      40076
Track 37      41318
Track 38      42473
Track 39      43708
Track 40      44855
Track 41      46136
Track 42      47206
Track 43      48274
Track 44      49385
Track 45      50400
Track 46      51403
Track 47      52327
Track 48      53311
Track 49      54310
Track 50      55247


### Para filtrar, possível percorrer todos os registros onde os valores não são nulos até a coluna "Track 49" 

In [10]:
registros_sem_nulos = new_df.loc[:, :'Track 66'].dropna()

In [11]:
len(registros_sem_nulos)

40356

### Como é possível ver, agora temos 40356 registros sem valores nulos, o que melhora a eficácia do modelo

In [12]:
pd.set_option('display.max_rows', None)
registros_sem_nulos.isna().sum()

Track 1     0
Track 2     0
Track 3     0
Track 4     0
Track 5     0
Track 6     0
Track 7     0
Track 8     0
Track 9     0
Track 10    0
Track 11    0
Track 12    0
Track 13    0
Track 14    0
Track 15    0
Track 16    0
Track 17    0
Track 18    0
Track 19    0
Track 20    0
Track 21    0
Track 22    0
Track 23    0
Track 24    0
Track 25    0
Track 26    0
Track 27    0
Track 28    0
Track 29    0
Track 30    0
Track 31    0
Track 32    0
Track 33    0
Track 34    0
Track 35    0
Track 36    0
Track 37    0
Track 38    0
Track 39    0
Track 40    0
Track 41    0
Track 42    0
Track 43    0
Track 44    0
Track 45    0
Track 46    0
Track 47    0
Track 48    0
Track 49    0
Track 50    0
Track 51    0
Track 52    0
Track 53    0
Track 54    0
Track 55    0
Track 56    0
Track 57    0
Track 58    0
Track 59    0
Track 60    0
Track 61    0
Track 62    0
Track 63    0
Track 64    0
Track 65    0
Track 66    0
dtype: int64

### Agora que temos os dados tratados, o próximo passo é converter em uma matriz, pois é assim que o modelo consegue ler os dados.

In [13]:
registros = []

num_colunas = 66

for linha in range(0, len(registros_sem_nulos)):
    registro_linha = []
    for coluna in range(0, num_colunas):
        registro_linha.append(registros_sem_nulos.iloc[linha, coluna])
    registros.append(registro_linha)

### Para facilitar as próximas execuções, essa matriz será salva em um arquivo para posteriormente ser lida novamente.

In [14]:
with open('matrix.txt', 'w') as arquivo:
    json.dump(registros, arquivo)

In [ ]:
import pandas as pd
import os
import numpy as np
import json
import random

# Ler a lista de volta do arquivo JSON
with open('matrix.txt', 'r') as arquivo:
    matriz = json.load(arquivo)

In [ ]:
matriz[7][9]

'Hospital / Right On'

### Agora com a matriz em mãos, é possível executar o modelo importado da lib apyori e editar os parâmetros

### - Definindo valores de suporte, confiança e Lift:

In [ ]:
support = 0.0030
# Quero que uma associação occora pelo menos 150 vezes 
# na base de dados para entrar na regra, então uso esse valor para chegar em um valor aproximado

confidence = 0.005
# A medida de confiança define a probabilidade de ocorrência do consequente na transacao,
# dado que a transacao já contém os antecedentes. Porém, se o item consequente é muito popular na base de dados, ele irá 
# aparecer na maioria das transações, o que pode levar a uma análise incorreta, para corrigir isso, o LIFT é utilizado.

lift = 5
# O valor mais importante, geralmente é definido que é válido analisar associações onde esse valor é maior que 1.

#Explicação completa desses valores aqui -> https://towardsdatascience.com/association-rules-2-aa9a77241654

In [ ]:
from apyori import apriori

arules = apriori(matriz,min_support=support, min_confidence=confidence,min_lift=lift)

association_results = list(arules)

len(association_results)

16130

In [ ]:
def frozenset_encoder(obj):
    if isinstance(obj, frozenset):
        return list(obj)
    raise TypeError(repr(obj) + " is not JSON serializable")

In [ ]:
with open("arules.txt", "w") as file:
    json.dump(association_results, file, default=frozenset_encoder)

In [252]:
import pandas as pd
import os
import numpy as np
import json
import random

with open("arules.txt",'r') as arquivo_leitura:
    association_results = json.load(arquivo_leitura)

In [253]:
# len(association_results)
association_results[0]

[['*NSYNC / Bye Bye Bye', "*NSYNC / It's Gonna Be Me"],
 0.003568242640499554,
 [[['*NSYNC / Bye Bye Bye'],
   ["*NSYNC / It's Gonna Be Me"],
   0.2691588785046729,
   53.773147034329604],
  [["*NSYNC / It's Gonna Be Me"],
   ['*NSYNC / Bye Bye Bye'],
   0.7128712871287128,
   53.7731470343296]]]

In [254]:
import pandas as pd
import os
import numpy as np
import json
import random

results = []
# conjuntos_filtrados = [conjunto for conjunto in association_results if '' not in conjunto]

for item in association_results:
    
    pair = item[0]
    items = [x for x in pair]
    
    value6 = items[0]+" & "+items[1]
    value0 = items[0]
    value1 = items[1]
    
    value2 = item[1]
    
    value3 = item[2][0][2]
    value4 = item[2][0][3]
    
    rows = (value6,value0,value1,value2,value3,value4)
    results.append(rows)
    
labels = ["Associação","Musica 1","Musica 2", "Support", "Confidence","Lift"]

music_suggestion = pd.DataFrame.from_records(results,columns=labels)

In [255]:
len(music_suggestion)

16130

In [262]:
music_suggestion.drop_duplicates(subset=['Musica 1', 'Musica 2'], inplace=True)

In [263]:
music_suggestion.sort_values(by='Lift',ascending=False)

Associação  \
9692   Michael Bublé / White Christmas (Duet With Sha...   
10753  Michael Bublé / I'll Be Home For Christmas & M...   
10920  Michael Bublé / Blue Christmas & Michael Bublé...   
10913  Michael Bublé / Blue Christmas & Michael Bublé...   
10755  Michael Bublé / Blue Christmas & Michael Bublé...   
...                                                  ...   
734           Drake / Controlla & Big Sean / Bounce Back   
1322   J. Cole / Wet Dreamz & Chance The Rapper / No ...   
2283               Drake / Jumpman & Flo Rida / My House   
4955   Waka Flocka Flame / No Hands (feat. Roscoe Das...   
5433        Rob $tone / Chill Bill & Rihanna / Needed Me   

                                                Musica 1  \
9692   Michael Bublé / White Christmas (Duet With Sha...   
10753         Michael Bublé / I'll Be Home For Christmas   
10920                     Michael Bublé / Blue Christmas   
10913                     Michael Bublé / Blue Christmas   
10755                     Michael Bublé / Blue Christmas   
...                                                  ...   
734                                    Drake / Controlla   
1322                                J. Cole / Wet Dreamz   
2283                                     Drake / Jumpman   
4955   Waka Flocka Flame / No Hands (feat. Roscoe Das...   
5433                              Rob $tone / Chill Bill   

                                                Musica 2   Support  \
9692                        Michael Bublé / Silent Night  0.003197   
10753                       Michael Bublé / Silent Night  0.003197   
10920  Michael Bublé / It's Beginning To Look A Lot L...  0.003271   
10913  Michael Bublé / Jingle Bells (feat. The Puppin...  0.003197   
10755  Michael Bublé / Christmas (Baby Please Come Home)  0.003172   
...                                                  ...       ...   
734                               Big Sean / Bounce Back  0.003890   
1322   Chance The Rapper / No Problem (feat. Lil Wayn...  0.003841   
2283                                 Flo Rida / My House  0.004336   
4955         Migos / Bad and Boujee (feat. Lil Uzi Vert)  0.004138   
5433                                 Rihanna / Needed Me  0.003444   

       Confidence        Lift  
9692     0.791411  236.579141  
10753    0.791411  233.125431  
10920    0.809816  231.779663  
10913    0.791411  231.436116  
10755    0.785276  231.318257  
...           ...         ...  
734      0.115356    5.011110  
1322     0.117960    5.010959  
2283     0.118805    5.009928  
4955     0.117193    5.004698  
5433     0.128111    5.000030  

[6110 rows x 6 columns]

In [258]:
import plotly.express as px


df = music_suggestion.sort_values(by='Lift',ascending=False)

fig = px.scatter_3d(df, x='Support', y='Confidence', z='Lift',hover_name='Associação',color='Lift',color_continuous_scale='hot')

fig.show()

![image.png](attachment:image.png)

### Exemplo, se uma pessoa ouviu a música X, quais músicas indicar para ela adicionar na playlist?

In [290]:
sorted_suggestions = music_suggestion.sort_values(by='Lift',ascending=False)

In [291]:
sorted_suggestions[sorted_suggestions['Associação'].str.contains("Smells Like Teen Spirit")]

Associação  \
5130  Nirvana / Lithium & Nirvana / Smells Like Teen...   
5127  Nirvana / Come As You Are & Nirvana / Smells L...   
5129  Nirvana / Smells Like Teen Spirit & Nirvana / ...   
4838  Metallica / Enter Sandman & Nirvana / Smells L...   
3077  Foo Fighters / Everlong & Nirvana / Smells Lik...   
5133  Red Hot Chili Peppers / Californication & Nirv...   
5134  Red Hot Chili Peppers / Under The Bridge & Nir...   
5136  blink-182 / All The Small Things & Nirvana / S...   
3332  Nirvana / Smells Like Teen Spirit & Guns N' Ro...   
5131  Oasis / Wonderwall - Remastered & Nirvana / Sm...   
5135  Nirvana / Smells Like Teen Spirit & Sublime / ...   
5132  Nirvana / Smells Like Teen Spirit & Queen / Bo...   

                                      Musica 1  \
5130                         Nirvana / Lithium   
5127                 Nirvana / Come As You Are   
5129         Nirvana / Smells Like Teen Spirit   
4838                 Metallica / Enter Sandman   
3077                   Foo Fighters / Everlong   
5133   Red Hot Chili Peppers / Californication   
5134  Red Hot Chili Peppers / Under The Bridge   
5136          blink-182 / All The Small Things   
3332         Nirvana / Smells Like Teen Spirit   
5131           Oasis / Wonderwall - Remastered   
5135         Nirvana / Smells Like Teen Spirit   
5132         Nirvana / Smells Like Teen Spirit   

                                         Musica 2   Support  Confidence  \
5130            Nirvana / Smells Like Teen Spirit  0.004510    0.666667   
5127            Nirvana / Smells Like Teen Spirit  0.006294    0.610577   
5129                   Nirvana / Heart-Shaped Box  0.004014    0.593407   
4838            Nirvana / Smells Like Teen Spirit  0.003197    0.300699   
3077            Nirvana / Smells Like Teen Spirit  0.003048    0.300000   
5133            Nirvana / Smells Like Teen Spirit  0.003915    0.220670   
5134            Nirvana / Smells Like Teen Spirit  0.004163    0.234637   
5136            Nirvana / Smells Like Teen Spirit  0.003618    0.203911   
3332          Guns N' Roses / Sweet Child O' Mine  0.003692    0.177170   
5131            Nirvana / Smells Like Teen Spirit  0.003271    0.184358   
5135                           Sublime / Santeria  0.003048    0.171788   
5132  Queen / Bohemian Rhapsody - Remastered 2011  0.003122    0.175978   

           Lift  
5130  37.575419  
5127  34.414026  
5129  33.446252  
4838  16.948353  
3077  16.908939  
5133  13.492991  
5134  13.280513  
5136  10.456184  
3332   9.985857  
5131   8.910099  
5135   8.454469  
5132   6.606283

### Com as associações acima, é possível perceber que uma pessoa que gosta do gênero com certeza iria curtir algumas dessas recomendações.

### Outro exemplo, se a pessoa ouviu o artista X, quais artistas ela pode gostar?

In [292]:
def split_string(input_string,part):
    # Count occurrences of "/"
    count_slash = input_string.count("/")
    # If there's more than one occurrence
    if  "AC / DC" in input_string:
        # Find the position of the second occurrence of "/"
        second_slash_index = input_string.find("/", input_string.find("/") + 1)
        # Split at the position of the second occurrence
        return "/".join(input_string.split("/", 2)[:2])
    else:
        # Split normally
        return input_string.split("/")[part].strip()

In [298]:
split_string("AC / DC / ...",0)

'AC / DC '

In [294]:
df_artistas_sorted = sorted_suggestions
df_artistas_sorted['Artista 1'] = df_artistas_sorted['Musica 1'].apply(lambda x: split_string(x, 0))
df_artistas_sorted['Artista 2'] = df_artistas_sorted['Musica 2'].apply(lambda x: split_string(x, 0))
df_artistas_sorted['Musica 1'] = df_artistas_sorted['Musica 1'].apply(lambda x: split_string(x, 1))
df_artistas_sorted['Musica 2'] = df_artistas_sorted['Musica 2'].apply(lambda x: split_string(x, 1))

df_artistas_sorted = df_artistas_sorted[['Associação', 'Musica 1', 'Musica 2', 'Artista 1', 'Artista 2', 'Support','Confidence','Lift']] 

df_artistas_sorted

Associação  \
9692   Michael Bublé / White Christmas (Duet With Sha...   
10753  Michael Bublé / I'll Be Home For Christmas & M...   
10920  Michael Bublé / Blue Christmas & Michael Bublé...   
10913  Michael Bublé / Blue Christmas & Michael Bublé...   
10755  Michael Bublé / Blue Christmas & Michael Bublé...   
...                                                  ...   
734           Drake / Controlla & Big Sean / Bounce Back   
1322   J. Cole / Wet Dreamz & Chance The Rapper / No ...   
2283               Drake / Jumpman & Flo Rida / My House   
4955   Waka Flocka Flame / No Hands (feat. Roscoe Das...   
5433        Rob $tone / Chill Bill & Rihanna / Needed Me   

                                                Musica 1  \
9692            White Christmas (Duet With Shania Twain)   
10753                         I'll Be Home For Christmas   
10920                                     Blue Christmas   
10913                                     Blue Christmas   
10755                                     Blue Christmas   
...                                                  ...   
734                                            Controlla   
1322                                          Wet Dreamz   
2283                                             Jumpman   
4955   No Hands (feat. Roscoe Dash and Wale) - Explic...   
5433                                          Chill Bill   

                                          Musica 2          Artista 1  \
9692                                  Silent Night      Michael Bublé   
10753                                 Silent Night      Michael Bublé   
10920  It's Beginning To Look A Lot Like Christmas      Michael Bublé   
10913     Jingle Bells (feat. The Puppini Sisters)      Michael Bublé   
10755            Christmas (Baby Please Come Home)      Michael Bublé   
...                                            ...                ...   
734                                    Bounce Back              Drake   
1322       No Problem (feat. Lil Wayne & 2 Chainz)            J. Cole   
2283                                      My House              Drake   
4955           Bad and Boujee (feat. Lil Uzi Vert)  Waka Flocka Flame   
5433                                     Needed Me          Rob $tone   

               Artista 2   Support  Confidence        Lift  
9692       Michael Bublé  0.003197    0.791411  236.579141  
10753      Michael Bublé  0.003197    0.791411  233.125431  
10920      Michael Bublé  0.003271    0.809816  231.779663  
10913      Michael Bublé  0.003197    0.791411  231.436116  
10755      Michael Bublé  0.003172    0.785276  231.318257  
...                  ...       ...         ...         ...  
734             Big Sean  0.003890    0.115356    5.011110  
1322   Chance The Rapper  0.003841    0.117960    5.010959  
2283            Flo Rida  0.004336    0.118805    5.009928  
4955               Migos  0.004138    0.117193    5.004698  
5433             Rihanna  0.003444    0.128111    5.000030  

[6110 rows x 8 columns]

In [295]:
def busca_artistas(artista):
    df_matches = df_artistas_sorted[(df_artistas_sorted['Artista 1'] == artista) | (df_artistas_sorted['Artista 2'] == artista)]

    # split_by_slash = lambda x: x.split("/")[0]
    df_matches['Recomendacoes'] = df_matches['Artista 1'] + df_matches['Artista 2']
    df_artistas =  df_matches['Recomendacoes']
    return df_artistas

In [296]:
artista = "Guns N' Roses"
x=busca_artistas(artista)

# removed_searched_band = lambda x, artist: [item for item in x if item != artist]

artists_list_filtered = [artist.replace(artista, '').strip() for artist in x]

artists_list_filtered = list(filter(None, artists_list_filtered))
artists_list_filtered = list(set(artists_list_filtered))

artists_list_filtered

C:\Users\pcozz\AppData\Local\Temp\ipykernel_14608\1935377541.py:5: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



['Lynyrd Skynyrd',
 'Def Leppard',
 'Kansas',
 'Nirvana',
 'Aerosmith',
 'Van Halen',
 'Neil Diamond',
 'Bryan Adams',
 'Ozzy Osbourne',
 'AC',
 'Metallica',
 'KISS',
 'Boston',
 'Journey',
 'Queen',
 'Billy Joel',
 'The Rolling Stones',
 'The Animals',
 'Creedence Clearwater Revival',
 'Led Zeppelin',
 'Rick Springfield',
 'Van Morrison',
 'Eagles',
 'Survivor',
 'Bon Jovi',
 'Elton John']

In [297]:
print(artists_list_filtered)

['Lynyrd Skynyrd', 'Def Leppard', 'Kansas', 'Nirvana', 'Aerosmith', 'Van Halen', 'Neil Diamond', 'Bryan Adams', 'Ozzy Osbourne', 'AC', 'Metallica', 'KISS', 'Boston', 'Journey', 'Queen', 'Billy Joel', 'The Rolling Stones', 'The Animals', 'Creedence Clearwater Revival', 'Led Zeppelin', 'Rick Springfield', 'Van Morrison', 'Eagles', 'Survivor', 'Bon Jovi', 'Elton John']
